src/core/qgsmaplayerstylemanager.h

```diff
@@ -195,4 +194,47 @@ class CORE_EXPORT QgsMapLayerStyleManager : public QObject
     QString defaultStyleName() const;
 };
 
+
+/**
+ * \ingroup core
+ * Restore overridden layer style on destruction.
+ *
+ * \since QGIS 3.2
+ */
+class CORE_EXPORT QgsLayerStyleOverride
+{
+  public:
+
+    /**
+     * Construct a style override object associated with a map layer.
+     * The overridden style will be restored upon object destruction.
+     */
+    QgsLayerStyleOverride( QgsMapLayer *layer )
+      : mLayer( layer )
+    {
+    }
+
+    ~QgsLayerStyleOverride()
+    {
+      if ( mLayer && mStyleOverridden )
+        mLayer->styleManager()->restoreOverrideStyle();
+    }
+
+    /**
+     * Temporarily apply a different style to the layer. The argument
+     * can be either a style name or a full QML style definition.
+     */
+    void setOverrideStyle( const QString &style )
+    {
+      if ( mLayer && mStyleOverridden )
+        mLayer->styleManager()->restoreOverrideStyle();
+      mLayer->styleManager()->setOverrideStyle( style );
```

Comment:
if we allow mLayer to be nullptr, then this line should be also enclosed in "if (mLayer)" statement?